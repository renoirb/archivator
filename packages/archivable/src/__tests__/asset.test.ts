/* eslint-env jest */

import {
  createHashFunction,
  getHashes,
  CryptoCommonHashingFunctions,
} from '../crypto'
import {
  assetReferenceHandlerFactory,
  createNormalizedAssetReferenceMap,
  DocumentAssets,
  extractNormalizedAssetDest,
  assetFileExtensionNormalizer,
  NormalizedAsset,
} from '..'

import { fixtures } from '.'

/**
 * The string "http://www.example.org/a/b/c.png" encoded in Hexadecimal, hashing as...
 *
 * {@link CryptoCommonHashingFunctions}
 */
type ExampleSameReferenceHashes = Extract<
  'sha1' | 'mdc2' | 'md5' | 'sha512' | 'sha224',
  CryptoCommonHashingFunctions
>
const sameReferenceDifferentHash: Record<ExampleSameReferenceHashes, string> = {
  sha1: '4c49ccbf4cdbdbcfc7f91cf87f6e9636008e4a97.png',
  mdc2: '0c6a93f9d825332c4a9a74e31730690c.png',
  md5: '6a324cd1a0e4e480c4db3e0558360527.png',
  sha512:
    '038cfb9ef4324be8ef490dc2fb3b1807f775afff693202bf7823e842aa5168612a441132bac051f90f6fcb59942324826d5885370e4dbe8465baa915aa1b6382.png',
  sha224: '6291f0c9dac71cbbb9ebbfe09a560c204aff4e9cbdd913eda3dfef21.png',
}

describe('NormalizedAsset', () => {
  test('Happy-Path', () => {
    // Just so one can toggle this once, for sanity.
    const localTestHashName: ExampleSameReferenceHashes = 'md5'

    /**
     * Imagine we have an HTML document with URL http://www.example.org/a/
     * With an image tag `<img src="b/c.png" />` (our "asset", below)
     *
     * Which should make 'b/c.png' from 'http://www.example.org/a/' to be 'http://www.example.org/a/b/c.png'
     * That's what "src" should refer to.
     */
    const sourceDocument = 'http://www.example.org/a/'
    const match = 'b/c.png'
    const src = 'http://www.example.org/a/b/c.png' // sourceDocument + match

    const subject = new NormalizedAsset(sourceDocument, match)
    expect(subject).toMatchObject({ src, match })
    expect(subject).toMatchSnapshot()

    // BEGIN Separate concerns of reference from NormalizedAssetType
    const referenceHandler = assetReferenceHandlerFactory(
      createHashFunction(localTestHashName, 'hex'),
      assetFileExtensionNormalizer,
    )
    // END Separate concerns...

    const extractedAssetReference = referenceHandler(subject)
    expect(extractedAssetReference).toMatchSnapshot()
    expect(extractedAssetReference).toMatchObject({
      reference: sameReferenceDifferentHash[localTestHashName],
    })

    expect(() => extractNormalizedAssetDest(subject, sourceDocument)).toThrow(
      'Missing asset reference, make sure you’ve used assetReferenceHandlerFactory before using extractNormalizedAssetDest.',
    )

    /**
     * Make sure above throws ^ so we can avoid bugs later own below
     * Separate concerns of where to store file from NormalizedAssetType.
     *
     * extractNormalizedAssetDest works on merged object, since we're separating concerns
     * from NormalizedAsset and what should be the reference and dest value.
     * We’re merging it manually here.
     */
    const manuallyMergedSubject = Object.assign(
      {},
      { ...subject, ...extractedAssetReference },
    )
    expect(manuallyMergedSubject).toHaveProperty(
      'reference',
      sameReferenceDifferentHash[localTestHashName],
    )

    const extractedAssetDest = extractNormalizedAssetDest(
      manuallyMergedSubject,
      sourceDocument,
    )
    expect(extractedAssetDest).toMatchObject({
      dest: 'example.org/a/' + sameReferenceDifferentHash[localTestHashName],
    })

    // console.log('run', {
    //   sourceDocument,
    //   match,
    //   src,
    //   reference,
    //   asset: subject,
    // })
  })

  const hash = 'sha1'
  test.each(fixtures.loadNormalizedAssetAlphaJsonFixture())(
    `%s:\n\thash:\t\t\t${hash}\n\tsrc:\t\t\t%s\n\texpected:\t\t%s\n\treason:\t\t\t%s\n\tduration:\t\t`,
    (sourceDocument, src, expected) => {
      const subject = new NormalizedAsset(sourceDocument, src)
      // console.log('assets.json run', {sourceDocument, src, expected, reason: _, asset: subject})
      expect(subject).toMatchSnapshot()
      expect(subject).toHaveProperty('src', expected)
      expect(subject).toHaveProperty('match', src)
      // Those below should be handled separately
      expect(subject).toHaveProperty('reference', null)
      expect(subject).toHaveProperty('dest', null)
    },
  )

  const resourceUrl = 'http://www.example.org/a/b/c.png'
  // @ts-ignore
  test.skip.each(getHashes())(`${resourceUrl} hashed in %s`, (hash) => {
    const hashed = createHashFunction(hash, 'hex')(resourceUrl)
    // expect(hashed).toMatchSnapshot()
    console.log('getHashes run', { resourceUrl, hash, hashed })
  })
})

describe('DocumentAssets', () => {
  test('Happy-Path', () => {
    const sourceDocument = 'http://renoirboulanger.com/page/3/'
    const matches = [
      // Case 1: Fully qualified URL that is local to the site
      'http://renoirboulanger.com/wp-content/themes/twentyseventeen/assets/images/header.jpg',
      // Case 2: Fully qualified URL that is outside
      'https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png',
      // Case 3: Fully qualified  URL that is outside and protocol relative
      '//www.gravatar.com/avatar/cbf8c9036c204fe85e15155f9d70faec?s=500',
      // Case 4: Relative URL to the domain name, starting at root
      '/wp-content/themes/renoirb/assets/img/zce_logo.jpg',
      // Case 5: Relative URL to the current source document
      '../../avatar.jpg',
    ]
    const collection = new DocumentAssets(sourceDocument, matches)
    // @ts-ignore
    const arrayified = [...collection]
    expect(arrayified).toMatchSnapshot()
  })

  test('Documentation examples', () => {
    const sourceDocument =
      'http://blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path'
    const matches = ['//www.example.org/a/b/c.png']
    const subject = new DocumentAssets(sourceDocument, matches)
    // @ts-ignore
    const arrayified = [...subject]
    expect(arrayified).toMatchSnapshot()
    // Just sanity-check, since we **DO NOT CALL** DocumentAssets.setReferenceHandler(), it should default sha1 in hex
    expect(arrayified[0]).toHaveProperty(
      'reference',
      sameReferenceDifferentHash.sha1,
    )
  })

  test('setReferenceHandler', () => {
    const sourceDocument =
      'http://blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path'
    const matches = ['//www.example.org/a/b/c.png']
    const subject = new DocumentAssets(sourceDocument, matches)

    // BEGIN Separate concerns of reference from NormalizedAssetType
    const referenceHandler = assetReferenceHandlerFactory(
      createHashFunction('md5', 'hex'),
      assetFileExtensionNormalizer,
    )
    subject.setReferenceHandler(referenceHandler)
    // END BEGIN Separate concerns...

    // @ts-ignore
    const arrayified = [...subject]
    expect(arrayified).toMatchSnapshot()
    // Just sanity-check, since we **DO CALL** DocumentAssets.setReferenceHandler()
    expect(arrayified[0]).toHaveProperty(
      'reference',
      sameReferenceDifferentHash.md5,
    )
  })
})

describe('createNormalizedAssetReferenceMap', () => {
  test('Happy-Path', () => {
    const sourceDocument = 'http://renoirboulanger.com/page/3/'
    const matches = [
      // Case 1: Fully qualified URL that is local to the site
      'http://renoirboulanger.com/wp-content/themes/twentyseventeen/assets/images/header.jpg',
      // Case 2: Fully qualified URL that is outside
      'https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png',
      // Case 3: Fully qualified  URL that is outside and protocol relative
      '//www.gravatar.com/avatar/cbf8c9036c204fe85e15155f9d70faec?s=500',
      // Case 4: Relative URL to the domain name, starting at root
      '/wp-content/themes/renoirb/assets/img/zce_logo.jpg',
      // Case 5: Relative URL to the current source document
      '../../avatar.jpg',
    ]
    const collection = new DocumentAssets(sourceDocument, matches)
    const subject = createNormalizedAssetReferenceMap(collection)
    expect(subject).toMatchSnapshot()
  })
})
