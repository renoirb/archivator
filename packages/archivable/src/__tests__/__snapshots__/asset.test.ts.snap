// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DocumentAssets Documentation examples 1`] = `
Array [
  Object {
    "dest": "blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path/4c49ccbf4cdbdbcfc7f91cf87f6e9636008e4a97.png",
    "match": "//www.example.org/a/b/c.png",
    "reference": "4c49ccbf4cdbdbcfc7f91cf87f6e9636008e4a97.png",
    "src": "http://www.example.org/a/b/c.png",
  },
]
`;

exports[`DocumentAssets Happy-Path 1`] = `
Array [
  Object {
    "dest": "renoirboulanger.com/page/3/430e2156af17010e0d8ffcd726a95595fa71a4fd.jpg",
    "match": "http://renoirboulanger.com/wp-content/themes/twentyseventeen/assets/images/header.jpg",
    "reference": "430e2156af17010e0d8ffcd726a95595fa71a4fd.jpg",
    "src": "http://renoirboulanger.com/wp-content/themes/twentyseventeen/assets/images/header.jpg",
  },
  Object {
    "dest": "renoirboulanger.com/page/3/b41de0a18bbb0871b22e0f5c466b3cd2f498807d.png",
    "match": "https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png",
    "reference": "b41de0a18bbb0871b22e0f5c466b3cd2f498807d.png",
    "src": "https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png",
  },
  Object {
    "dest": "renoirboulanger.com/page/3/63dc122dfd3c702e12714fbe4ba744e463c49edb",
    "match": "//www.gravatar.com/avatar/cbf8c9036c204fe85e15155f9d70faec?s=500",
    "reference": "63dc122dfd3c702e12714fbe4ba744e463c49edb",
    "src": "http://www.gravatar.com/avatar/cbf8c9036c204fe85e15155f9d70faec?s=500",
  },
  Object {
    "dest": "renoirboulanger.com/page/3/840257d7de220958ca4cc05a3c0ee337e2b0401d.jpg",
    "match": "/wp-content/themes/renoirb/assets/img/zce_logo.jpg",
    "reference": "840257d7de220958ca4cc05a3c0ee337e2b0401d.jpg",
    "src": "http://renoirboulanger.com/wp-content/themes/renoirb/assets/img/zce_logo.jpg",
  },
  Object {
    "dest": "renoirboulanger.com/page/3/37fd63a34f42ed3b012b9baac82e97fbe9f9c067.jpg",
    "match": "../../avatar.jpg",
    "reference": "37fd63a34f42ed3b012b9baac82e97fbe9f9c067.jpg",
    "src": "http://renoirboulanger.com/avatar.jpg",
  },
]
`;

exports[`DocumentAssets setReferenceHandler 1`] = `
Array [
  Object {
    "dest": "blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path/6a324cd1a0e4e480c4db3e0558360527.png",
    "match": "//www.example.org/a/b/c.png",
    "reference": "6a324cd1a0e4e480c4db3e0558360527.png",
    "src": "http://www.example.org/a/b/c.png",
  },
]
`;

exports[`NormalizedAsset Happy-Path 1`] = `
Object {
  "dest": null,
  "match": "b/c.png",
  "reference": null,
  "src": "http://www.example.org/a/b/c.png",
}
`;

exports[`NormalizedAsset Happy-Path 2`] = `
Object {
  "hasExtension": true,
  "reference": "6a324cd1a0e4e480c4db3e0558360527.png",
}
`;

exports[`NormalizedAsset http://blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path:
	hash:			sha1
	src:			//example.org/a/b.png
	expected:		http://example.org/a/b.png
	reason:			Ignore document origin if resource has full URL, protocol relative, non TLS
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "//example.org/a/b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path:
	hash:			sha1
	src:			http://example.org/a/b.png
	expected:		http://example.org/a/b.png
	reason:			Ignore document origin if resource has full URL
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "http://example.org/a/b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://elsewhere.org/a/b:
	hash:			sha1
	src:			//example.org/foo/bar.jpg
	expected:		http://example.org/foo/bar.jpg
	reason:			Refering document is NOT over TLS, but referenced document is over protocol relative. Enforce same protocol
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "//example.org/foo/bar.jpg",
  "reference": null,
  "src": "http://example.org/foo/bar.jpg",
}
`;

exports[`NormalizedAsset http://example.org/:
	hash:			sha1
	src:			./a.png
	expected:		http://example.org/a.png
	reason:			Explicit parent from current context root
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "./a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/:
	hash:			sha1
	src:			A.PNG
	expected:		http://example.org/A.PNG
	reason:			Case sensitive
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "A.PNG",
  "reference": null,
  "src": "http://example.org/A.PNG",
}
`;

exports[`NormalizedAsset http://example.org/:
	hash:			sha1
	src:			a.png
	expected:		http://example.org/a.png
	reason:			Relative path when at root
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/:
	hash:			sha1
	src:			a.png?foo=BAR
	expected:		http://example.org/a.png?foo=BAR
	reason:			Does not remove search query string
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "a.png?foo=BAR",
  "reference": null,
  "src": "http://example.org/a.png?foo=BAR",
}
`;

exports[`NormalizedAsset http://example.org/:
	hash:			sha1
	src:			a.png?foo=BAR&BaZ=Bizz
	expected:		http://example.org/a.png?foo=BAR&BaZ=Bizz
	reason:			Does not remove search query string with amperstands
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "a.png?foo=BAR&BaZ=Bizz",
  "reference": null,
  "src": "http://example.org/a.png?foo=BAR&BaZ=Bizz",
}
`;

exports[`NormalizedAsset http://example.org/:
	hash:			sha1
	src:			a/b
	expected:		http://example.org/a/b
	reason:			Do not affect if no file extension
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "a/b",
  "reference": null,
  "src": "http://example.org/a/b",
}
`;

exports[`NormalizedAsset http://example.org/a.html:
	hash:			sha1
	src:			b.png
	expected:		http://example.org/b.png
	reason:			Relative path when using file name with .html at top directory
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b.png",
  "reference": null,
  "src": "http://example.org/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a/:
	hash:			sha1
	src:			./b.png
	expected:		http://example.org/a/b.png
	reason:			Explicit parent from current context root
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "./b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a/:
	hash:			sha1
	src:			b.png
	expected:		http://example.org/a/b.png
	reason:			Relative path from a sub folder
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a/:
	hash:			sha1
	src:			b/c.png
	expected:		http://example.org/a/b/c.png
	reason:			Deeper relative path
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b/c.png",
  "reference": null,
  "src": "http://example.org/a/b/c.png",
}
`;

exports[`NormalizedAsset http://example.org/a/c.do:
	hash:			sha1
	src:			b.png
	expected:		http://example.org/a/b.png
	reason:			Document is one folder deep and file with .do extension, asset relative to same directory where file is
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a/c.html:
	hash:			sha1
	src:			b.png
	expected:		http://example.org/a/b.png
	reason:			Relative path when using file name with .html
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a/c/:
	hash:			sha1
	src:			../b.png
	expected:		http://example.org/a/b.png
	reason:			Relative path going into 1 parent up
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "../b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a/c:
	hash:			sha1
	src:			../b.png
	expected:		http://example.org/a/b.png
	reason:			Relative path going into 1 parent up, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "../b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a:
	hash:			sha1
	src:			./b.png
	expected:		http://example.org/a/b.png
	reason:			Explicit parent from current context root in sub folder, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "./b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a:
	hash:			sha1
	src:			B.pnG
	expected:		http://example.org/a/B.pnG
	reason:			Case sensitive, relative path from a sub folder, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "B.pnG",
  "reference": null,
  "src": "http://example.org/a/B.pnG",
}
`;

exports[`NormalizedAsset http://example.org/a:
	hash:			sha1
	src:			Something%20We%20Should%20Not%20Do%20But%20Yet%20Is%20Done.jpg
	expected:		http://example.org/a/Something%20We%20Should%20Not%20Do%20But%20Yet%20Is%20Done.jpg
	reason:			Do not touch spaces and case
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "Something%20We%20Should%20Not%20Do%20But%20Yet%20Is%20Done.jpg",
  "reference": null,
  "src": "http://example.org/a/Something%20We%20Should%20Not%20Do%20But%20Yet%20Is%20Done.jpg",
}
`;

exports[`NormalizedAsset http://example.org/a:
	hash:			sha1
	src:			b(WTH).png
	expected:		http://example.org/a/b(WTH).png
	reason:			Parens are not valid, but some uses them, must support them
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b(WTH).png",
  "reference": null,
  "src": "http://example.org/a/b(WTH).png",
}
`;

exports[`NormalizedAsset http://example.org/a:
	hash:			sha1
	src:			b.png
	expected:		http://example.org/a/b.png
	reason:			Relative path from a sub folder, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset http://example.org/a:
	hash:			sha1
	src:			b/c.png
	expected:		http://example.org/a/b/c.png
	reason:			Deeper relative path, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "b/c.png",
  "reference": null,
  "src": "http://example.org/a/b/c.png",
}
`;

exports[`NormalizedAsset http://example.org/b/c.do:
	hash:			sha1
	src:			/a.png
	expected:		http://example.org/a.png
	reason:			Document is one folder deep and file with .do extension, asset is referring at the root
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "/a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/b/c.html:
	hash:			sha1
	src:			../../../../../a.png
	expected:		http://example.org/a.png
	reason:			File is HTML document, not a directory and asset is relative and goes beyond root directory. Force at root.
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "../../../../../a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/b/c.html:
	hash:			sha1
	src:			../a.png
	expected:		http://example.org/a.png
	reason:			Document is one folder deep with .do file extension, asset refers to 1 parent directory up
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "../a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/b/c.html:
	hash:			sha1
	src:			/a.png
	expected:		http://example.org/a.png
	reason:			Document is one folder deep and file with .html extension, asset is referring at the root
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "/a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/b/c:
	hash:			sha1
	src:			../../../a.png
	expected:		http://example.org/a.png
	reason:			Asset is relative and goes beyond root directory
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "../../../a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/bogus/:
	hash:			sha1
	src:			/a.png
	expected:		http://example.org/a.png
	reason:			Absolute path, without trailing slash, ignoring sub folder from hostname
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "/a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/bogus/deep/path:
	hash:			sha1
	src:			/a.png
	expected:		http://example.org/a.png
	reason:			Absolute path, without trailing slash, ignoring sub folder from hostname
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "/a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/bogus:
	hash:			sha1
	src:			/a.png
	expected:		http://example.org/a.png
	reason:			Absolute path, ignoring sub folder from hostname
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "/a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org/foo:
	hash:			sha1
	src:			1*23.png
	expected:		http://example.org/foo/1*23.png
	reason:			Star is not typical in path, must support them
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "1*23.png",
  "reference": null,
  "src": "http://example.org/foo/1*23.png",
}
`;

exports[`NormalizedAsset http://example.org/ignored/also_ignored/:
	hash:			sha1
	src:			../../a.jpg
	expected:		http://example.org/a.jpg
	reason:			Relative path going into 2 parents up
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "../../a.jpg",
  "reference": null,
  "src": "http://example.org/a.jpg",
}
`;

exports[`NormalizedAsset http://example.org/ignored/also_ignored/and_too:
	hash:			sha1
	src:			../../../a.jpg
	expected:		http://example.org/a.jpg
	reason:			Relative path going into 3 parents up, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "../../../a.jpg",
  "reference": null,
  "src": "http://example.org/a.jpg",
}
`;

exports[`NormalizedAsset http://example.org:
	hash:			sha1
	src:			./a.png
	expected:		http://example.org/a.png
	reason:			Explicit parent from current context root, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "./a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset http://example.org:
	hash:			sha1
	src:			a.png
	expected:		http://example.org/a.png
	reason:			Relative path when at root, without trailing slash
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "a.png",
  "reference": null,
  "src": "http://example.org/a.png",
}
`;

exports[`NormalizedAsset https://blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path:
	hash:			sha1
	src:			http://example.org/a/b.png
	expected:		http://example.org/a/b.png
	reason:			Ignore document origin if resource has full URL, even over TLS, we shall take asset resource URL as-is
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "http://example.org/a/b.png",
  "reference": null,
  "src": "http://example.org/a/b.png",
}
`;

exports[`NormalizedAsset https://blackhole.webpagetest.org/renoirb/archivator/test/normalizer/assets/ignore-path-tls:
	hash:			sha1
	src:			//example.org/a/b.png
	expected:		https://example.org/a/b.png
	reason:			Ignore document origin if resource has full URL, protocol relative, over TLS
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "//example.org/a/b.png",
  "reference": null,
  "src": "https://example.org/a/b.png",
}
`;

exports[`NormalizedAsset https://elsewhere.org/a/b:
	hash:			sha1
	src:			//example.org/foo/bar.jpg
	expected:		https://example.org/foo/bar.jpg
	reason:			Refering document is over TLS, but referenced document is over protocol relative. Enforce same protocol
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "//example.org/foo/bar.jpg",
  "reference": null,
  "src": "https://example.org/foo/bar.jpg",
}
`;

exports[`NormalizedAsset https://example.org/:
	hash:			sha1
	src:			a.png
	expected:		https://example.org/a.png
	reason:			Maintains protocol
	duration:		 1`] = `
Object {
  "dest": null,
  "match": "a.png",
  "reference": null,
  "src": "https://example.org/a.png",
}
`;

exports[`createNormalizedAssetReferenceMap Happy-Path 1`] = `
Map {
  "http://renoirboulanger.com/wp-content/themes/twentyseventeen/assets/images/header.jpg" => "430e2156af17010e0d8ffcd726a95595fa71a4fd.jpg",
  "https://s3.amazonaws.com/github/ribbons/forkme_right_gray_6d6d6d.png" => "b41de0a18bbb0871b22e0f5c466b3cd2f498807d.png",
  "//www.gravatar.com/avatar/cbf8c9036c204fe85e15155f9d70faec?s=500" => "63dc122dfd3c702e12714fbe4ba744e463c49edb",
  "/wp-content/themes/renoirb/assets/img/zce_logo.jpg" => "840257d7de220958ca4cc05a3c0ee337e2b0401d.jpg",
  "../../avatar.jpg" => "37fd63a34f42ed3b012b9baac82e97fbe9f9c067.jpg",
}
`;
